#### Submission Deadline: 31/05/2023  -  23:59

# Question: 01

**MAP REDUCE - DATASET:** write map reduce program(s) using python-mr in colab notebook and complete the following exercises

  - Find the files which contain the following phrases
    - “draping the folds of her velvet dress”
    - “grim clasp to the iron stanchions”
    - “Lo, I teach you the Superman!”
    - “Through that divine allegiance”
  - Find the file which has maximum number of tokens
  - Find the file which has the longest sentence (a sentence ends in a full stop and can span multiple lines)


**Dataset:** 

  - [pg1513.txt](https://www.gutenberg.org/cache/epub/1513/pg1513.txt)<br>
  - [2160-0.txt](https://www.gutenberg.org/files/2160/2160-0.txt)<br>
  - [pg11.txt](https://www.gutenberg.org/cache/epub/11/pg11.txt)<br>
  - [pg2600.txt](https://www.gutenberg.org/cache/epub/2600/pg2600.txt)<br>
  - [345.txt.utf-8](https://www.gutenberg.org/ebooks/345.txt.utf-8)<br>
  - [64317.txt.utf-8](https://www.gutenberg.org/ebooks/64317.txt.utf-8)<br>
  - [70815.txt.utf-8](https://www.gutenberg.org/ebooks/70815.txt.utf-8)<br>
  - [1080-0.txt](https://www.gutenberg.org/files/1080/1080-0.txt)<br>
  - [70817.txt.utf-8](https://www.gutenberg.org/ebooks/70817.txt.utf-8)<br>
  - [1998.txt.utf-8](https://www.gutenberg.org/ebooks/1998.txt.utf-8)<br>



## Answer: 01

#### Supporting URL
#### https://mrjob.readthedocs.io/en/latest/

#### Importing common library 

In [11]:
import os
import urllib.request

In [12]:
workingPath = 'C:/Users/Admin/Coding-M-Tech/AI Lab 2 - ID5003W/Assignment 1'

In [13]:
!pip install --quiet mrjob

#### Changing the working directory 

In [14]:
os.chdir(workingPath)

#### Common Variable

In [15]:
## All the downloadable url
url_pg1513 = 'https://www.gutenberg.org/cache/epub/1513/pg1513.txt'
url_2160 = 'https://www.gutenberg.org/files/2160/2160-0.txt'
url_pg11 = 'https://www.gutenberg.org/cache/epub/11/pg11.txt'
url_pg2600 = 'https://www.gutenberg.org/cache/epub/2600/pg2600.txt'
url_pg345 = 'https://www.gutenberg.org/cache/epub/345/pg345.txt'
url_pg64317 = 'https://www.gutenberg.org/cache/epub/64317/pg64317.txt'
url_pg70815 = 'https://www.gutenberg.org/cache/epub/70815/pg70815.txt'
url_1080 = 'https://www.gutenberg.org/files/1080/1080-0.txt'
url_pg70817 = 'https://www.gutenberg.org/cache/epub/70817/pg70817.txt'
url_pg1998 = 'https://www.gutenberg.org/cache/epub/1998/pg1998.txt'

#### Map reduce Functionality 

###### Phrase Searcher

In [16]:
%%file phrase_search.py
from mrjob.job import MRJob
from mrjob.step import MRStep
import os

class FindPhrase(MRJob):
     
    def configure_args(self):
        super(FindPhrase, self).configure_args()
        self.add_passthru_arg('--phrase', type=str, help='Phrase to find in files')

    def mapper_init(self):
        self.phrase = self.options.phrase

    def mapper(self, _, line):
        filename = os.getenv('mapreduce_map_input_file')
        file_name = filename.split('/')[-1]
        if self.phrase in line:
            yield (self.phrase,file_name), 1

    def reducer(self, key, values): 
        yield key, f"{sum(values)} occurance of : {key[0]} phrase, is present in : {key[1]} file"

if __name__ == '__main__':
    FindPhrase.run()


Overwriting phrase_search.py


#### Max Token Searcher

In [17]:
%%file max_token_search.py
from mrjob.job import MRJob
from mrjob.step import MRStep
import os

class FindMaxToken(MRJob):
     
    def mapper(self, _, line):
        filename = os.getenv('mapreduce_map_input_file')
        tokens = line.strip().split()
        file_name = filename.split('/')[-1]
        yield file_name, len(tokens)

    def reducer(self, key, values):
        yield key, f" file has token value of : {sum(values) }"

if __name__ == '__main__':
    FindMaxToken.run()



Overwriting max_token_search.py


#### Maximum Length Sentence

In [18]:
%%file max_length_search.py
from mrjob.job import MRJob
from mrjob.step import MRStep
import os
import re

class FindMaxSentenceLength(MRJob):
     
    def mapper(self, _, line):
        filename = os.getenv('mapreduce_map_input_file')
        sentences = re.split(r'(?<=[.])\s+', line)
        file_name = filename.split('/')[-1]
        for sentence in sentences:
            yield file_name, len(sentence)

    def reducer(self, key, values):
        yield key, f" file has sentence of maximum {max(values)} length"

if __name__ == '__main__':
    FindMaxSentenceLength.run()

Overwriting max_length_search.py


#### Creating a dataset folder

In [19]:
!mkdir Q1-Dataset

A subdirectory or file Q1-Dataset already exists.


#### Downloading the file

In [20]:
urllib.request.urlretrieve(url_pg1513,'Q1-Dataset/pg1513.txt')
urllib.request.urlretrieve(url_2160,'Q1-Dataset/2160-0.txt')
urllib.request.urlretrieve(url_pg11,'Q1-Dataset/pg11.txt')
urllib.request.urlretrieve(url_pg2600,'Q1-Dataset/pg2600.txt')
urllib.request.urlretrieve(url_pg345,'Q1-Dataset/pg345.txt')
urllib.request.urlretrieve(url_pg64317,'Q1-Dataset/pg64317.txt')
urllib.request.urlretrieve(url_pg70815,'Q1-Dataset/pg70815.txt')
urllib.request.urlretrieve(url_1080,'Q1-Dataset/1080-0.txt')
urllib.request.urlretrieve(url_pg70817,'Q1-Dataset/pg70817.txt')
urllib.request.urlretrieve(url_pg1998,'Q1-Dataset/pg1998.txt')

('Q1-Dataset/pg1998.txt', <http.client.HTTPMessage at 0x2a6bc0cbd00>)

##### 1a. Running Phrase Finder method

In [21]:
!python phrase_search.py --phrase="draping the folds of her velvet dress" "Q1-Dataset/"

!python phrase_search.py --phrase="grim clasp to the iron stanchions" "Q1-Dataset/"

!python phrase_search.py --phrase="Lo, I teach you the Superman!" "Q1-Dataset/"

!python phrase_search.py --phrase="Through that divine allegiance" "Q1-Dataset/"

No configs found; falling back on auto-configuration
No configs specified for inline runner
Creating temp directory C:\Users\Admin\AppData\Local\Temp\phrase_search.Admin.20230527.144457.412711
Running step 1 of 1...
job output is in C:\Users\Admin\AppData\Local\Temp\phrase_search.Admin.20230527.144457.412711\output
Streaming final output from C:\Users\Admin\AppData\Local\Temp\phrase_search.Admin.20230527.144457.412711\output...
Removing temp directory C:\Users\Admin\AppData\Local\Temp\phrase_search.Admin.20230527.144457.412711...


["draping the folds of her velvet dress","pg2600.txt"]	"1 occurance of : draping the folds of her velvet dress phrase, is present in : pg2600.txt file"
["grim clasp to the iron stanchions","pg345.txt"]	"1 occurance of : grim clasp to the iron stanchions phrase, is present in : pg345.txt file"


No configs found; falling back on auto-configuration
No configs specified for inline runner
Creating temp directory C:\Users\Admin\AppData\Local\Temp\phrase_search.Admin.20230527.144459.316365
Running step 1 of 1...
job output is in C:\Users\Admin\AppData\Local\Temp\phrase_search.Admin.20230527.144459.316365\output
Streaming final output from C:\Users\Admin\AppData\Local\Temp\phrase_search.Admin.20230527.144459.316365\output...
Removing temp directory C:\Users\Admin\AppData\Local\Temp\phrase_search.Admin.20230527.144459.316365...


["Lo, I teach you the Superman!","pg1998.txt"]	"1 occurance of : Lo, I teach you the Superman! phrase, is present in : pg1998.txt file"


No configs found; falling back on auto-configuration
No configs specified for inline runner
Creating temp directory C:\Users\Admin\AppData\Local\Temp\phrase_search.Admin.20230527.144501.191245
Running step 1 of 1...
job output is in C:\Users\Admin\AppData\Local\Temp\phrase_search.Admin.20230527.144501.191245\output
Streaming final output from C:\Users\Admin\AppData\Local\Temp\phrase_search.Admin.20230527.144501.191245\output...
Removing temp directory C:\Users\Admin\AppData\Local\Temp\phrase_search.Admin.20230527.144501.191245...
No configs found; falling back on auto-configuration
No configs specified for inline runner
Creating temp directory C:\Users\Admin\AppData\Local\Temp\phrase_search.Admin.20230527.144503.058394
Running step 1 of 1...
job output is in C:\Users\Admin\AppData\Local\Temp\phrase_search.Admin.20230527.144503.058394\output
Streaming final output from C:\Users\Admin\AppData\Local\Temp\phrase_search.Admin.20230527.144503.058394\output...
Removing temp directory C:\Users

### Conclusion 1a .
#### "draping the folds of her velvet dress" phrase is available in pg2600.txt file
#### "grim clasp to the iron stanchions"   phrase is available in pg345.txt file
#### "Lo, I teach you the Superman!"  phrase is available in 1998.txt file

##### 1b. Running Max Token Finder method

In [22]:
!python max_token_search.py "Q1-Dataset/*"

"Q1-Dataset\\1080-0.txt"	" file has token value of : 6512"
"Q1-Dataset\\2160-0.txt"	" file has token value of : 151379"
"Q1-Dataset\\pg11.txt"	" file has token value of : 29590"
"Q1-Dataset\\pg1513.txt"	" file has token value of : 29002"
"Q1-Dataset\\pg1998.txt"	" file has token value of : 113997"
"Q1-Dataset\\pg2600.txt"	" file has token value of : 566330"
"Q1-Dataset\\pg345.txt"	" file has token value of : 164459"
"Q1-Dataset\\pg64317.txt"	" file has token value of : 51257"
"Q1-Dataset\\pg70815.txt"	" file has token value of : 82782"
"Q1-Dataset\\pg70817.txt"	" file has token value of : 28184"


No configs found; falling back on auto-configuration
No configs specified for inline runner
Creating temp directory C:\Users\Admin\AppData\Local\Temp\max_token_search.Admin.20230527.144522.041254
Running step 1 of 1...
job output is in C:\Users\Admin\AppData\Local\Temp\max_token_search.Admin.20230527.144522.041254\output
Streaming final output from C:\Users\Admin\AppData\Local\Temp\max_token_search.Admin.20230527.144522.041254\output...
Removing temp directory C:\Users\Admin\AppData\Local\Temp\max_token_search.Admin.20230527.144522.041254...


### Conclusion 1b. 
#### pg2600.txt has maximum number of token i.e. 566330

##### 1c. Running Find Max Sentence Length method

In [23]:
!python max_length_search.py "Q1-Dataset/*"

"Q1-Dataset\\1080-0.txt"	" file has sentence of maximum 72 length"
"Q1-Dataset\\2160-0.txt"	" file has sentence of maximum 79 length"
"Q1-Dataset\\pg11.txt"	" file has sentence of maximum 82 length"
"Q1-Dataset\\pg1513.txt"	" file has sentence of maximum 90 length"
"Q1-Dataset\\pg1998.txt"	" file has sentence of maximum 78 length"
"Q1-Dataset\\pg2600.txt"	" file has sentence of maximum 74 length"
"Q1-Dataset\\pg345.txt"	" file has sentence of maximum 73 length"
"Q1-Dataset\\pg64317.txt"	" file has sentence of maximum 72 length"
"Q1-Dataset\\pg70815.txt"	" file has sentence of maximum 72 length"
"Q1-Dataset\\pg70817.txt"	" file has sentence of maximum 72 length"


No configs found; falling back on auto-configuration
No configs specified for inline runner
Creating temp directory C:\Users\Admin\AppData\Local\Temp\max_length_search.Admin.20230527.144529.496349
Running step 1 of 1...
job output is in C:\Users\Admin\AppData\Local\Temp\max_length_search.Admin.20230527.144529.496349\output
Streaming final output from C:\Users\Admin\AppData\Local\Temp\max_length_search.Admin.20230527.144529.496349\output...
Removing temp directory C:\Users\Admin\AppData\Local\Temp\max_length_search.Admin.20230527.144529.496349...


### Conclusion 1c. 
#### pg1513.txt has maximum sentence length of 90.

# Conclusion : 

### Q. Find the files which contain the following phrases
### A. "draping the folds of her velvet dress" phrase is available in pg2600.txt file
### "grim clasp to the iron stanchions"   phrase is available in pg345.txt file
### "Lo, I teach you the Superman!"  phrase is available in 1998.txt file


### Q. Find the file which has maximum number of tokens
### A. pg2600.txt has maximum number of token i.e. 566330


### Q. Find the file which has the longest sentence (a sentence ends in a full stop and can span multiple lines)
### A. pg1513.txt has maximum sentence length of 90.


# Question: 02

**Write a PySpark RDD program to compute:**


  - The list of common (overlapping / intersection ) words across all the ten files listed above.
  - List the word frequency of common words across files as a dict ( tip: use join method in the rdd). The size of the dict equals the list of common words and each key in the top level dict contains the word count corresponding to all the ten files

    - {“word”: {“pg1513.txt”:10,”2160-0.txt”:25,... } 


## Answer: 02

In [24]:
!pip install pyspark

In [26]:
from pyspark import SparkConf, SparkContext
from operator import add

# Initialize Spark
conf = SparkConf().setMaster('local').setAppName('CommonWords')
sc = SparkContext.getOrCreate(conf=conf)

# List of files
files = ['Q1-Dataset/pg1513.txt','Q1-Dataset/2160-0.txt','Q1-Dataset/pg11.txt','Q1-Dataset/pg2600.txt','Q1-Dataset/pg345.txt','Q1-Dataset/pg64317.txt','Q1-Dataset/pg70815.txt','Q1-Dataset/1080-0.txt','Q1-Dataset/pg70817.txt','Q1-Dataset/pg1998.txt']
# Process each file to get RDDs of words
rdd_list = []
for file in files:
    rdd = sc.textFile(file)
    words = rdd.flatMap(lambda line: line.split(" "))
    rdd_list.append(words)

# Compute intersection of all files to get common words
common_words_rdd = rdd_list[0]
for rdd in rdd_list[1:]:
    common_words_rdd = common_words_rdd.intersection(rdd)

common_words = common_words_rdd.distinct().collect()

print("Common Words are:", common_words)

# Now process each file to get word frequencies
freq_dict = {}
for file in files:
    rdd = sc.textFile(file)
    words = rdd.flatMap(lambda line: line.split(" "))
    word_freq = words.countByValue()
    for word in common_words:
        if word in word_freq:
            if word not in freq_dict:
                freq_dict[word] = {}
            freq_dict[word][file] = word_freq[word]

print(freq_dict)
sc.stop()

Py4JJavaError: An error occurred while calling z:org.apache.spark.api.python.PythonRDD.collectAndServe.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 3 in stage 3.0 failed 1 times, most recent failure: Lost task 3.0 in stage 3.0 (TID 5) (Mou-Aloy.mshome.net executor driver): java.net.SocketException: Connection reset by peer: socket write error
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(Unknown Source)
	at java.net.SocketOutputStream.write(Unknown Source)
	at java.io.BufferedOutputStream.flushBuffer(Unknown Source)
	at java.io.BufferedOutputStream.flush(Unknown Source)
	at java.io.DataOutputStream.flush(Unknown Source)
	at org.apache.spark.api.python.BasePythonRunner$WriterThread.$anonfun$run$1(PythonRunner.scala:443)
	at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:2088)
	at org.apache.spark.api.python.BasePythonRunner$WriterThread.run(PythonRunner.scala:274)

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2785)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2721)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2720)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2720)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1206)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1206)
	at scala.Option.foreach(Option.scala:407)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1206)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2984)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2923)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2912)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:971)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2263)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2284)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2303)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2328)
	at org.apache.spark.rdd.RDD.$anonfun$collect$1(RDD.scala:1019)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:405)
	at org.apache.spark.rdd.RDD.collect(RDD.scala:1018)
	at org.apache.spark.api.python.PythonRDD$.collectAndServe(PythonRDD.scala:193)
	at org.apache.spark.api.python.PythonRDD.collectAndServe(PythonRDD.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:374)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
	at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
	at java.lang.Thread.run(Unknown Source)
Caused by: java.net.SocketException: Connection reset by peer: socket write error
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(Unknown Source)
	at java.net.SocketOutputStream.write(Unknown Source)
	at java.io.BufferedOutputStream.flushBuffer(Unknown Source)
	at java.io.BufferedOutputStream.flush(Unknown Source)
	at java.io.DataOutputStream.flush(Unknown Source)
	at org.apache.spark.api.python.BasePythonRunner$WriterThread.$anonfun$run$1(PythonRunner.scala:443)
	at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:2088)
	at org.apache.spark.api.python.BasePythonRunner$WriterThread.run(PythonRunner.scala:274)


# Question: 03

Perform an exploratory data analysis on the primary_data.csv of the Mushroom dataset [Secondary Mushroom Dataset](https://archive.ics.uci.edu/ml/datasets/Secondary+Mushroom+Dataset) using pandas and answer the following questions:
  - How many mushrooms are poisonous?
  - Which family contains most poisonous mushrooms?
  - Do all poisonous mushrooms have a ring?
  - Which seasons do poisonous mushrooms grow typically?

## Answer: 03

#### Importing Libraries

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os

#### Common Variables

In [ ]:
working_dir_path = 'C:/Users/Admin/Coding-M-Tech/AI Lab 2 - ID5003W/Assignment 1/Q3-Dataset'
primaryFileName = 'primary_data.csv'

#### Change the folder location 

In [ ]:
os.chdir(working_dir_path)

#### Loading primary data

In [ ]:
df_primary_data = pd.read_csv(primaryFileName,sep=';')

In [ ]:
df_primary_data

#### Perform basic statistical analysis and missing value check

In [ ]:
df_primary_data.describe(include='object')

In [ ]:
df_primary_data.info()

In [ ]:
df_primary_data.isna().sum().sort_values(ascending=False)

In [ ]:
poisonous_mushroom = df_primary_data[df_primary_data['class'] == 'p']
edible_mushroom = df_primary_data[df_primary_data['class'] == 'e']

##### Graphical representation of the distribution of poisonous and edible mushroom

In [ ]:
print('==========Distribution of poisonous and edible mushroom============')
mushroom_type_count= df_primary_data['class'].value_counts()  
sns.barplot(x=mushroom_type_count.index, y=mushroom_type_count.values)
plt.ylabel('Number of data', fontsize=9)
plt.xlabel('class', fontsize=9)
plt.xticks(rotation=90)
plt.show();
df_primary_data['class'].value_counts()

### Q. How many mushrooms are poisonous?
### A. 96 mushrooms are poisonous and 77 are edible

##### Graphical representation of the distribution of poisonous mushroom among different family

In [ ]:
poisonous_mushroom_type_count= poisonous_mushroom['family'].value_counts()  
sns.barplot(x=poisonous_mushroom_type_count.index, y=poisonous_mushroom_type_count.values)
plt.ylabel('Number of data', fontsize=9)
plt.xlabel('Family', fontsize=9)
plt.xticks(rotation=90)
plt.show();
df_primary_data['family'].value_counts()

### Q. Which family contains most poisonous mushrooms?
### A. Tricholoma, mushroom family has highest number of poisonous mushrooms.

##### Analysis of poisonous mushroom based on its ring characteristics

In [ ]:
np.unique(poisonous_mushroom['has-ring'])

##### It has been observed that the unique value of mushroom has-ring column is [f] & [t], which clearly indicate that only ring exists among the mushroom which has 'has-ring' column value as [t], so we will filter that out and get the desired result.

In [ ]:
ring_poisonous_mushroom = poisonous_mushroom[poisonous_mushroom['has-ring'] != '[t]']

In [ ]:
ring_poisonous_mushroom

### Q. Do all poisonous mushrooms have a ring?
### A. No, all the poisonous mushroom does not have ring, only 70 out of 96 poisonous mushroom have ring.

##### Analysis of poisonous mushroom based on the season 

In [ ]:
np.unique(poisonous_mushroom['season'])

##### Graphical representation of the distribution of poisonous mushroom based on it season of growth

In [ ]:
poisonous_mushroom_type_count= poisonous_mushroom['season'].value_counts()  
sns.barplot(x=poisonous_mushroom_type_count.index, y=poisonous_mushroom_type_count.values)
plt.ylabel('Number of data', fontsize=9)
plt.xlabel('Season', fontsize=9)
plt.xticks(rotation=90)
plt.show();
df_primary_data['season'].value_counts()

### Q. Which seasons do poisonous mushrooms grow typically?
### A. Over Summer and Autumn most of the poisonous mushrooms typically grows.

# Conclusion : 

### Q. How many mushrooms are poisonous?
### A. 96 mushrooms are poisonous and 77 are edible
### Q. Which family contains most poisonous mushrooms?
### A. Tricholoma, mushroom family has highest number of poisonous mushrooms.
### Q. Do all poisonous mushrooms have a ring?
### A. No, all the poisonous mushroom does not have ring, only 70 out of 96 poisonous mushroom have ring.
### Q. Which seasons do poisonous mushrooms grow typically?
### A. Over Summer and Autumn most of the poisonous mushrooms typically grows.